File: backend/src/Sites/StreamingCommunity.cs
using System.Linq;
using System.Net;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Stronzflix.Media;
using Stronzflix.MediaPlayers;
using Stronzflix.Utils;

namespace Stronzflix.Sites
{
    public class StreamingCommunity : Site
    {
        private readonly string search_url;
        private readonly Dictionary<string, string> inhertia;

        private readonly string cdn;

        public StreamingCommunity(string url)
            : base("StreamingCommunity", url)
        {
            this.search_url = base.Url + "/search?q=";
            this.cdn = string.Join("//cdn.", url.Split("//"));
            string inhertia_version = this.GetInertia();
            this.inhertia = new Dictionary<string, string>()
            {
                { "X-Inertia", "true" },
                { "X-Inertia-Version", inhertia_version },
            };
        }

        private string GetInertia()
        {
            string html = SimpleHTTP.Get(base.Url);
            string match = SimpleRegex.Search(@"version&quot;:&quot;(?<inertia>[a-z0-9]+)&quot;", html);
            return match;
        }

        public override Result[] Search(string query)
        {
            string json = SimpleHTTP.Get(this.search_url + query, this.inhertia);
            Dictionary<string, object> data = JsonConvert.DeserializeObject<Dictionary<string, object>>(json);
            JArray titles = (JArray)((JObject)data["props"])["titles"];

            List<Result> results = new List<Result>();
            foreach (JObject title in titles.Cast<JObject>())
            {
                string title_id = (string)title["id"];
                string title_slug = (string)title["slug"];
                JArray images = (JArray)title["images"];
                string poster = "";
                foreach(JObject image in images.Cast<JObject>())
                    if((string)image["type"] == "poster")
                    {
                        poster = (string)image["filename"];
                        break;
                    }

                string title_url = "/titles/" + title_id + "-" + title_slug;
                string title_name = (string)title["name"];
                string poster_url = this.cdn + "/images/" + poster;

                results.Add(new Result(this, title_name, title_url, poster_url));
            }

            return results.ToArray();
        }

        private Episode[] GetEpisodes(string season_url)
        {
            string json = SimpleHTTP.Get(base.Url + season_url, this.inhertia);
            Dictionary<string, object> data = JsonConvert.DeserializeObject<Dictionary<string, object>>(json);
            
            JObject loaded_season = (JObject)((JObject)data["props"])["loadedSeason"];
            string title_id = (string)((JObject)((JObject)data["props"])["title"])["id"];
            
            JArray episodes = (JArray)loaded_season["episodes"];
            
            List<Episode> episodes_list = new List<Episode>();

            foreach(JObject episode in episodes.Cast<JObject>())
            {
                string episode_id = (string)episode["id"];

                string episode_url = "/watch/" + title_id + "?e=" + episode_id;
                string episode_name = (string)episode["name"];

                episodes_list.Add(new Episode(episode_name, episode_url));
            }

            return episodes_list.ToArray();
        }

        public override Title GetTitle(string url)
        {
            string json = SimpleHTTP.Get(base.Url + url, this.inhertia);
            Dictionary<string, object> data = JsonConvert.DeserializeObject<Dictionary<string, object>>(json);
            JArray seasons = (JArray)((JObject)((JObject)data["props"])["title"])["seasons"];
            string name = (string)((JObject)((JObject)data["props"])["title"])["name"];
            
            List<Episode[]> episodes = new List<Episode[]>();

            foreach(JObject season in seasons.Cast<JObject>())
            {
                string season_number = (string)season["number"];
                string season_url = url + "/stagione-" + season_number;

                episodes.Add(this.GetEpisodes(season_url));
            }

            return new Series(name, episodes.ToArray());
        }

        public override string GetSource(string url)
        {
            string title_id = SimpleRegex.Search(@"watch/(\d+)", url);
            string episode_id = SimpleRegex.Search(@"\?e=(\d+)", url);
            string iframeSrc = "/iframe/" + title_id + "?episode_id=" + episode_id;

            string iframe = SimpleHTTP.Get(base.Url + iframeSrc);
            string src = WebUtility.HtmlDecode(SimpleRegex.Search("src=\"(.+?)\"", iframe));

            return new VixCloud().GetSource(src);
        }
    }
}

File: backend/sites/streamingcommunity.js
import { Site } from './site.js';
import { default as request } from 'sync-request';
import { decode } from 'html-entities';

export class StreamingCommunity extends Site
{
    constructor(url)
    {
        super("StreamingCommunity", url);

        this.search_url = this.url + "/search?q=";
        this.cdn = url.split("//").join("//cdn.");

        const inhertia_version = this.getInertia();
        this.inhertia =  {
            "X-Inertia": true,
            "X-Inertia-Version": inhertia_version,
        };
    }

    getInertia()
    {
        const response = request('GET', this.url, {
            headers: this.inhertia,
        });

        const version = response.getBody().toString().match(/version&quot;:&quot;(?<inertia>[a-z0-9]+)&quot;/)[1];
        return version;
    }

    search(query)
    {
        const json = JSON.parse(request('GET', this.search_url + query, {
            headers: this.inhertia,
        }).getBody('utf8'));

        const titles = json.props.titles;

        let results = []
        for (const title of titles)
        {
            const poster = title.images.find(image => image.type == "poster").filename;

            results.push({
                "site": this.name,
                "title": title.name,
                "url": "/titles/" + title.id + "-" + title.slug,
                "poster": this.cdn + "/images/" + poster
            });
        }

        return { "results": results };
    }

    getEpisodes(seasonUrl)
    {
        const json = JSON.parse(request('GET', this.url + seasonUrl, {
            headers: this.inhertia,
        }).getBody('utf8'));

        const season = json.props.loadedSeason;
        const titleId = json.props.title.id;
        
        let episodes = [];
        for(const episode of season.episodes)
        {
            const cover = episode.images.find(image => image.type == "cover").filename;
            episodes.push({
                "url": "/watch/" + titleId + "?e=" + episode.id,
                "name": episode.name,
                "cover": this.cdn + "/images/" + cover
            });
        }

        return episodes;
    }

    getSeries(url, title)
    {
        let episodes = [];
        for (const season of title.seasons)
        {
            const seasonUrl = url + "/stagione-" + season.number;
            episodes.push(this.getEpisodes(seasonUrl));
        }

        return { "seasons": episodes };
    }

    getFilm(title)
    {
        return { "url": "/watch/" + title.id, "name": title.name };
    }

    getTitle(url)
    {
        const json = JSON.parse(request('GET', this.url + url, {
            headers: this.inhertia,
        }).getBody('utf8'));

        const title = json.props.title;

        const result = title.type == "tv" ? this.getSeries(url, title) : this.getFilm(title);
   
        return { "title": result, "site": this.name, };
    }

    parseVixCloud(url)
    {
        const data = request('GET', url).getBody('utf8');

        const playlistUrl = data.match(/url: '(.+?)'/)[1];    
        
        const jsonString = data.match(/params: ({(.|\n)+?}),/)[1].replace(/\'/g, '"').replace(/ /g, "").replace(/\n/g, "").replace(/\",}/g, "\"}");
        const json = JSON.parse(jsonString);
        
        let param = "";
        for(const key in json)
            param += key + "=" + json[key] + "&";
        param = param.substring(0, param.length - 1);

        const playlist = playlistUrl + "?" + param;
        const source = request('GET', playlist).getBody('utf8');
        return source;
    }

    getSource(url)
    {
        const titleId = url.match(/watch\/(\d+)/)[1];
        const episodeId = url.match(/\?e=(\d+)/)[1];
        const iframeSrc = "/iframe/" + titleId + "?episode_id=" + episodeId;

        const iframe = request('GET', this.url + iframeSrc, {
            headers: { "User-Agent": "Stronzflix" }
        }).getBody('utf8');
        const src = decode(iframe.match(/src="(.+?)"/)[1]);

        return this.parseVixCloud(src);
    }
}
